Refactor rules loading result

mstemm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. . Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area rules

/area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

When loading rules, return a structure with the load result and lists of errors/warnings + file locations.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
update: when loading rules, return a structure with the load result and lists of errors/warnings + file locations.
```

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [12efc10](https://github.com/falcosecurity/falco/commits/12efc10f02f68f73ac3f4d9521acab0eb3eff1fc) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2098#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

jasondellaluce commented:
/milestone 0.33.0

mstemm commented:
One follow-on change I'd like to make is to extend the context struct to point inside a condition string/output string. I think that might require some libs changes such as annotating ASTs with yaml file positions. I see that the parser object has a position but I think for multi-line strings the yaml parsing changes the position. Let me know if I'm wrong on that and I can work on merging the positions.

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs
- [9a007b1](https://github.com/falcosecurity/falco/commits/9a007b1e5fd8a6fe884274ad44c00814e8bca0de) Squash w Rule loader changes to support result objects

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/falco/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [cb1841a](https://github.com/falcosecurity/falco/commits/cb1841acfefe8531ff592334d78c9d340cbf8351) DO NOT MERGE: Remove required_engine_version from falco engine load_rules APIs

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

mstemm commented:
I cherry-picked the commit from https://github.com/falcosecurity/falco/pull/2096 and included it in this PR temporarily, just so I could get the build/test working. Once 2096 is merged, I'll rebase and drop the cherry-picked commit.

mstemm commented:
Here are some example outputs that use the result structure and print the contents in various ways:

Validating some files, no errors:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml
Thu Jun 30 08:32:49 2022: Falco version 0.0.0
Thu Jun 30 08:32:49 2022: Falco initialized with configuration file /etc/falco/falco.yaml
Thu Jun 30 08:32:49 2022: Validating rules file(s):
Thu Jun 30 08:32:49 2022:    /etc/falco/falco_rules.yaml
Thu Jun 30 08:32:49 2022:    /etc/falco/falco_rules.local.yaml
/etc/falco/falco_rules.yaml: Ok
/etc/falco/falco_rules.local.yaml: Ok
```

Validating some files, some errors + warnings. This version of the output is terse, which matches old behavior:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml
Thu Jun 30 08:35:19 2022: Falco version 0.0.0
Thu Jun 30 08:35:19 2022: Falco initialized with configuration file /etc/falco/falco.yaml
Thu Jun 30 08:35:19 2022: Validating rules file(s):
Thu Jun 30 08:35:19 2022:    /etc/falco/falco_rules.yaml
Thu Jun 30 08:35:19 2022:    /etc/falco/falco_rules.local.yaml
Error: /etc/falco/falco_rules.yaml: Ok
/etc/falco/falco_rules.local.yaml: Invalid
```

If you add `-v` you get full details on the errors + warnings:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml -v
Thu Jun 30 08:38:48 2022: Falco version 0.0.0
Thu Jun 30 08:38:48 2022: Falco initialized with configuration file /etc/falco/falco.yaml
Thu Jun 30 08:38:48 2022: Validating rules file(s):
Thu Jun 30 08:38:48 2022:    /etc/falco/falco_rules.yaml
Thu Jun 30 08:38:48 2022:    /etc/falco/falco_rules.local.yaml
/etc/falco/falco_rules.local.yaml: Invalid
1 Errors:
In rule 'My Invalid Rule': (/etc/falco/falco_rules.local.yaml:37:2)
    exception 'invalid exception': (/etc/falco/falco_rules.local.yaml:43:6)
------
    - name: invalid exception
      ^
------
FE_LOAD_ERR_VALIDATE (Error validating rule/macro/list/exception objects): 'not-a-field' is not a supported filter field

1 Warnings:
In rule 'My Warning Rule': (/etc/falco/falco_rules.local.yaml:31:2)
------
- rule: My Warning Rule
  ^
------
FE_LOAD_NO_EVTTYPE (Condition has no event-type restriction): Rule matches too many evt.type values. This has a significant performance penalty.

Error: /etc/falco/falco_rules.yaml: Ok
/etc/falco/falco_rules.local.yaml: Invalid
```

And if you add `-o json_output=True` you get a json version of the errors + warnings instead. The result is printed to stdout, so if you redirect stderr you get *only* the validation result. This matches older falco behavior:

```
$ ./userspace/falco/falco -V /etc/falco/falco_rules.yaml -V /etc/falco/falco_rules.local.yaml -v -o json_output=True 2>/dev/null | jq "."
{
  "falco_load_results": [
    {
      "errors": [],
      "name": "/etc/falco/falco_rules.yaml",
      "successful": true,
      "warnings": []
    },
    {
      "errors": [
        {
          "code": "FE_LOAD_ERR_VALIDATE",
          "codedesc": "Error validating rule/macro/list/exception objects",
          "context": {
            "locations": [
              {
                "item_name": "My Invalid Rule",
                "item_type": "rule",
                "position": {
                  "column": 2,
                  "filename": "/etc/falco/falco_rules.local.yaml",
                  "line": 37,
                  "offset": 1246
                }
              },
              {
                "item_name": "invalid exception",
                "item_type": "exception",
                "position": {
                  "column": 6,
                  "filename": "/etc/falco/falco_rules.local.yaml",
                  "line": 43,
                  "offset": 1374
                }
              }
            ],
            "snippet": "    - name: invalid exception\n      ^\n"
          },
          "message": "'not-a-field' is not a supported filter field"
        }
      ],
      "name": "/etc/falco/falco_rules.local.yaml",
      "successful": false,
      "warnings": [
        {
          "code": "FE_LOAD_NO_EVTTYPE",
          "codedesc": "A rule condition matches too many evt.type values. This has a significant performance penalty. Make the condition more specific by adding an evt.type field or further restricting the number of evt.type values in the condition.",
          "context": {
            "locations": [
              {
                "item_name": "My Warning Rule",
                "item_type": "rule",
                "position": {
                  "column": 2,
                  "filename": "/etc/falco/falco_rules.local.yaml",
                  "line": 31,
                  "offset": 1139
                }
              }
            ],
            "snippet": "- rule: My Warning Rule\n  ^\n"
          },
          "message": "Rule matches too many evt.type values. This has a significant performance penalty."
        }
      ]
    }
  ]
}
```

jasondellaluce commented on test/falco_tests.yaml:
```diff
@@ -20,22 +20,55 @@ trace_files: !mux
-    rules_warning: False
 
+  # The rules_events part of this test was mistakenly disabled as a
+  # part of the lua to c++ rules switch. The rule loading
```
For context, this regression comes from way before the C++ porting (removal of `sinsp_rule_utils.lua`): https://github.com/falcosecurity/falco/pull/1715

mstemm replied:
Oh, ok! I will update the comment.

jasondellaluce commented on test/falco_tests.yaml:
```diff
@@ -251,350 +284,292 @@ trace_files: !mux
+        item_name: some macro
+        code: FE_LOAD_ERR_VALIDATE
+        message: "Undefined macro 'foo' used in filter."
+    validate_warnings:
```
This probably connects to my other comment: here the warning is not meaningful. We should have failed at the first error.

mstemm replied:
Although I didn't want to tackle it in this PR, which is already quite big, I was hoping that someday rules loading would not always fail at the first error. The exception-based method is a good way to stop loading and unwind to a top-level location, but it will only show the first problem, not all of the potential problems in a set of rules.

With this version, the result contains any warnings up until the first error, or all warnings if there are no errors.

I think having the validate_warnings/validate_errors properties as a list will support a future change to return multiple errors/warnings if we decide to change the implementation. We'd just have to change the expected values for any tests that have multiple errors, or additional warnings after any errors.

Does that sound ok?

jasondellaluce replied:
I see your point. I agree that `validate_warnings` and `validate_errors` must be lists to support multiple warnings and errors in the future, however I think changing our "fail at first error" model might be too much for a single PR. I'd suggest keeping the failure condition as-is for now, and eventually change it in a separate PR in which we document the rationale. Note that this would probably introduce user-facing breaking changes, so we might be to be careful about it!

mstemm replied:
Yeah I agree on not changing the rules loading code, it will not strive to return all errors. But I think it's fine to keep the test code the same, right? That's only visible internally and end users won't see the test code.

jasondellaluce commented on test/falco_tests.yaml:
```diff
@@ -1025,13 +1000,6 @@ trace_files: !mux
       - open_12: 0
       - open_13: 0
 
-  list_append_failure:
```
Why is this removed? Same goes for `macro_append_failure` and `rule_append_failure`.

mstemm replied:
These test cases were actually duplicates--{list,rule}_append_failure were duplicates of invalid_{list,rule}_append_dangling, just without the -V argument. macro_append_failure was a duplicate of invalid_append_macro_dangling, just without the -V argument and using a different rules file.

The point of the tests was really to check the validation, so I kept the validation versions.

jasondellaluce commented on test/falco_tests.yaml:
```diff
@@ -20,22 +20,55 @@ trace_files: !mux
-      - leading_in_not_equals_at_evttype
-      - not_with_evttypes
-      - not_with_evttypes_addl
+    validate_rules_file: rules/falco_rules_warnings.yaml
```
Are these all of type `LOAD_NO_EVTTYPE`? Or will we need to change them once we bring back the other kinds of warnings we had in `sinsp_rule_utils.lua`?

mstemm replied:
Yeah, they all ultimately result in the same (single) warning LOAD_NO_EVTTYPE, but have different ways to write a condition that ultimately matches too many event types e.g. "evt.type!=execve", "not evt.type=execve", etc. There's a lot of overlap between these cases and the ones in test_filter_evttype_resolver.cpp. These tests are much older, before we even had unit tests, but I think they're still valuable to test the warning result paths.

jasondellaluce commented on test/falco_test.py:
```diff
@@ -376,7 +360,68 @@ def check_outputs(self):
+                for vres in vobj["falco_load_results"]:
+                    for warning in vres["warnings"]:
+                        if warning["code"] == warnobj["code"]:
+                            if ("message" in warnobj and warning["message"] == warnobj["message"]) or ("message_contains" in warnobj and warnobj["message_contains"] in warning["message"]):
```
So basically `message` and `message_contains` seem to be mandatory, right? Should we make it optional? Should we document this somewhere?

mstemm replied:
message_contains is only used for the test engine_version_mismatch, where the error message will be dependent on the current embedded falco engine version e.g. "Rules require engine version 9999999, but engine version is 14". If it's a big deal I could switch everything to be a substring match, but I thought the distinction was useful.

I believe all of the other error messages are static--they used to contain "in rule xxx" snippets but all of that info is in the context instead. The engine version is a strange dynamic thing that doesn't really fit into a context.

jasondellaluce reviewed (commented):
Thanks for this effort Mark! I tried my best to review this, left you some comments.

On top of that, I think we also need to change the documentation here: https://github.com/falcosecurity/falco/blob/35db0b4a24344bc6c24022555f7d8531ad925136/falco.yaml#L77

`json_output` now does more than just formatting rule outputs, so I think we should document it properly. Alternatively, we should further separate responsibilities and have a separate config field like `json_load_result`.

mstemm commented:
I created a companion docs PR https://github.com/falcosecurity/falco-website/pull/657 that updates the docs for json_output to reflect that it controls both alerts and rule loading/validation results. I think a single option is preferable so I updated the docs rather than split into two separate options.

(I also updated the comments in falco.yaml)

mstemm commented on userspace/engine/falco_load_result.h:
```diff
@@ -0,0 +1,88 @@
+class falco_load_result {
+public:
+
+	enum error_code {
```
Sure, that's fine. I guess the values are still namespaced within the class so there isn't any chance of conflicts.

mstemm commented on userspace/engine/falco_load_result.h:
```diff
@@ -0,0 +1,88 @@
+#include <nlohmann/json.hpp>
+
+// Represents the result of loading a rules file.
+class falco_load_result {
```
Sure, I think we should put it in a falco namespace though as load_result by itself is generic. That brings up the question of whether we should add a namespace to falco_engine as well? The engine doesn't have any namespace at the moment, which is why I didn't use one for falco_load_result.

I don't want to get carried away with changes, so how about this--in this PR I simply add a falco namespace around load_result, and once this is merged we add a top level namespace to falco_engine? Maybe with an additional rename of the falco_engine class, we can bikeshed about it in the follow-on PR.

mstemm commented on userspace/engine/falco_load_result.h:
```diff
@@ -0,0 +1,88 @@
+	// impact.
+	static const std::string& error_desc(error_code ec);
+
+	enum warning_code {
```
Yep, removed in both.

mstemm commented on userspace/engine/rule_loader.h:
```diff
@@ -31,34 +33,108 @@ limitations under the License.
+		std::string rules_snippet;
+	};
+
+	class rule_load_exception : public std::exception {
```
Sure, fixed.

mstemm commented on userspace/engine/rule_loader.h:
```diff
@@ -31,34 +33,108 @@ limitations under the License.
+
+		// A chain of locations from the current item, its
+		// parent, possibly older ancestors.
+		std::list<location> m_locs;
```
I seem to be the only one who likes lists, I get this question whenever I use them :) I always choose lists unless I really want the O(1) random access. But everyone else likes vectors so I can change it.

mstemm commented on userspace/engine/rule_loader.h:
```diff
@@ -31,34 +33,108 @@ limitations under the License.
+
+		// Return a snippet of the provided rules content
+		// corresponding to this context.
+		std::string rules_snippet(const std::string& rules_content) const;
```
Yeah, I went back and forth on it, thinking that snippet by itself would be too generic. But now we have a vote for just snippet, so I'll change it everywhere.

mstemm commented on userspace/engine/rule_loader.cpp:
```diff
@@ -21,16 +21,399 @@ limitations under the License.
+		return "<No context available>\n";
+	}
+
+	rule_loader::context::location loc = m_locs.back();
```
Right, the other locations are used for the "In xxx (filename:line:col)" messages, but only the last location has a snippet.

I think as we use these messages we might find that it's helpful to have snippets from the other locations, but I think this is a good improvement (and is consistent) for now.

jasondellaluce replied:
Agree, thanks for the explanation!

mstemm commented on userspace/engine/rule_reader.cpp:
```diff
@@ -16,139 +16,167 @@ limitations under the License.
-		vector<YAML::Node>::iterator doc,
-		YAML::iterator node)
+template <typename T>
+static void decode_val(const YAML::Node& item, const char *key, T& out, const rule_loader::context& ctx, bool optional=false)
```
For around 70% of the calls to decode_val, optional is false, and I was trying to keep things clean by not specifying a value for the more common case. I'll add a separate function decode_optional_val() instead. I also made the other decode_xxx functions a bit more specific to what they extracted (e.g. decode_seq is now a helper w/ decode_tags/decode_items being the direct functions, decode_exception_info_entry is a helper with decode_exception_{fields,comps,values} being the direct functions. This sets optional in the right places for all these methods.

mstemm commented on userspace/engine/rule_reader.cpp:
```diff
@@ -222,57 +284,81 @@ static void read_item(
+			// have an enabled property. Use the enabled
+			// property to set the enabled status of an
+			// earlier rule.
+			if (!item["condition"].IsDefined() &&
```
I got stuck on this too, but I believe this logic is correct. I believe the goal here is that if a rule does not have any of condition/output/desc/priority, then it *must* have enabled, and the purpose is to simply set enabled for a previously defined rule with the same name.

The old code did this by setting has_defs to true only if all of condition/output/desc/priority were defined, and if !has_defs was true, requiring that enabled was present. I think that's a bit looser than the goal--if a rule defined some but not all of condition/output/desc/priority, has_defs would be false, !has_defs would be true, and enabled would be required. I don't think that's what we want, right?

The new code only checks enabled if *none* of condition/output/desc/priority are present. Otherwise, you jump past the if and try to extract each of condition/output/desc/priority individually (and throw exceptions if any are missing/can't be decoded).

Does this sound right to you too?

mstemm commented on userspace/engine/rule_reader.cpp:
```diff
@@ -222,57 +284,81 @@ static void read_item(
+		rule_loader::rule_info v(ctx);
+		v.name = name;
+
+		v.append = false;
```
Sure, that's fine. I initially thought that rule_reader simply parsed the objects from yaml into a list, with a separate pass for handling append. But since this version handles append while parsing the yaml, we can keep things simpler and have append just be a local variable.

mstemm commented on userspace/engine/rule_reader.cpp:
```diff
@@ -285,45 +371,40 @@ bool rule_reader::load(rule_loader::configuration& cfg, rule_loader& loader)
 			}
+			catch (rule_loader::rule_load_exception &e)
+			{
+				cfg.res->add_error(e.ec, e.msg, e.ctx, cfg.content);
```
Oh yes, although it should still iterate over each doc and return false if any have an error.

jasondellaluce replied:
I think this is related to failing at the first error in a ruleset file. However, I see your point here: different docs in the same file effectively represent different rulesets.

My take here is that I would return false here, and leave a todo. Then we can tackle this problem in another PR with a specific discussion. I think doing this kind of reasoning is too much for this contribution.

mstemm replied:
I do return false now, although it does iterate over the docs and returns false if any have an error. Do you mean this should return false immediately, and skip any remaining docs in the file?

The technique of multiple docs in a file isn't really that common--none of our rules files use it. So it's not like that distinction will be that noticeable.

jasondellaluce replied:
Thanks!

mstemm commented on userspace/engine/falco_load_result.h:
```diff
@@ -0,0 +1,88 @@
+	// This returns a short string with the success value and
+	// a list of errors/warnings. Suitable for simple one-line
+	// display.
+	virtual const std::string& as_summary() = 0;
```
Yeah, it's just verbosity. I think I'll still use separate functions for the actual implementation, but we can keep the "public" interface simple and just have a single as_string(). I think I'll actually use as_string(bool verbose) as "verbose" lines up with the command line option.

mstemm commented:
/retest

jasondellaluce reviewed (commented):
I agree on having a `falco::` namespace. This is also a first step towards uniformity because we have `falco::application` too. Would love to have `falco::engine` one day.

I advocate on maintaining the policy of failing at the first error in a ruleset for now. It is great to support having multiple errors in the test suite and in some parts of the code, but I think this PR already brings a big contribution on the table and this problem is out of context here. The issue of reporting more than one error for each ruleset doc is a broader discussion that would involve user-facing breaking changes. It's a great addition that we should keep in our roadmap, but my advice is to open a second PR just for that and have a discussion there.

mstemm commented:
> I agree on having a `falco::` namespace. This is also a first step towards uniformity because we have `falco::application` too. Would love to have `falco::engine` one day.
> 
> I advocate on maintaining the policy of failing at the first error in a ruleset for now. It is great to support having multiple errors in the test suite and in some parts of the code, but I think this PR already brings a big contribution on the table and this problem is out of context here. The issue of reporting more than one error for each ruleset doc is a broader discussion that would involve user-facing breaking changes. It's a great addition that we should keep in our roadmap, but my advice is to open a second PR just for that and have a discussion there.

Yep, I added a falco:: namespace to load_result. Further restructuring of namespaces can wait for a later PR.

And yes, returning multiple errors can also wait for a later PR. The current code will return multiple errors for the uncommon case of multiple docs in a single yaml file, unless you prefer I change it.

poiana commented:
LGTM label has been added.  <details>Git tree hash: 6fa1cdef4ca1b62680725be188d3a1a553cf5ff3</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2098#pullrequestreview-1059012868" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2098#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [jasondellaluce,mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

leogr reviewed (approved):
Hey @mstemm 

Thank for this PR. It's a significant improvement. 

Overall, it looks good to me. However, I've got a concern.
I don't see any real value in having non-verbose messages. The old behavior when loading rules was always to print verbose messages. It was advantageous. Think users that just copy/paste their Pod logs in a GitHub issue. Forcing them to redeploy Falco with `-v` would waste time.

That being said, I approve this PR anyway because I think it's valuable. Yet, I'd invite all @falcosecurity/falco-maintainers to reconsider verbosity mode when loading/validating rules. I believe verbosity should be enabled by default in those cases, or the previous behavior should be kept.

/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2098#pullrequestreview-1059012868" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2098#pullrequestreview-1061834680" title="Approved">leogr</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2098#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [jasondellaluce,leogr,mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

mstemm commented:
> Hey @mstemm
> 
> Thank for this PR. It's a significant improvement.
> 
> Overall, it looks good to me. However, I've got a concern. I don't see any real value in having non-verbose messages. The old behavior when loading rules was always to print verbose messages. It was advantageous. Think users that just copy/paste their Pod logs in a GitHub issue. Forcing them to redeploy Falco with `-v` would waste time.
> 
> That being said, I approve this PR anyway because I think it's valuable. Yet, I'd invite all @falcosecurity/falco-maintainers to reconsider verbosity mode when loading/validating rules. I believe verbosity should be enabled by default in those cases, or the previous behavior should be kept.
> 
> /approve

You're right about verbose--I really thought that the old behavior was to only print full details on the error when verbose was true. But I double-checked and 0.32.1 does print details both for `-r` and -`V`. So that's a regression and I'll change that in a follow-on (small) PR.

leogr commented:
> You're right about verbose--I really thought that the old behavior was to only print full details on the error when verbose was true. But I double-checked and 0.32.1 does print details both for `-r` and -`V`. So that's a regression and I'll change that in a follow-on (small) PR.

Thank you, Mark! :pray:
